core/arch/arm/kernel/vfp_a32.S

```diff
@@ -4,20 +4,30 @@
  */
 
 #include <asm.S>
+#include <arm32.h>
+#include <generated/arm32_sysreg.S>
 
 	.fpu	neon
 
 /* void vfp_save_extension_regs(uint64_t regs[VFP_NUM_REGS]); */
 FUNC vfp_save_extension_regs , :
-	vstm	r0!, {d0-d15}
-	vstm	r0, {d16-d31}
+	mov	r4, r0
```

Comment:
The calling convention requires `r4` to be preserved.
Why not keep `r0` as indexing register and use `r1` when reading CPACR?
Same comment applies below for `vfp_restore_extension_regs()`.

Reply:
According to ARM calling convention, `r0-r3` are used for function arguments/return values. `r4-r11` are used for local variables.
That is why I choose `r4` instead of `r1` to prevent function parameter extension.

Reply:
Not completely correct. Some quotes from section 5.1.1 Core registers below that table:

*The first four registers r0-r3 (a1-a4) are used to pass argument values into a subroutine and to return a result value from a function. They may also be used to hold intermediate values within a routine (but, in general, only between subroutine calls).*

*A subroutine must preserve the contents of the registers r4-r8, r10, r11 and SP (and r9 in PCS variants that designate r9 as v6).*

So you can do whatever you like with r0-r3, an eventual return value is ignored for this function as it's supposed to return `void`.

Reply:
Got your point. I'll revise it. Thanks for your suggestion.